src/main/scala/backends/futil/FutilBackend.scala

```diff
@@ -0,0 +1,304 @@
+package fuselang.backend.futil
+
+import fuselang.backend.futil.Futil._
+// import fuselang.backend.futil.Lib
+
+import fuselang.common._
+import Syntax._
+
+import Configuration._
+import CompilerError._
+
+private class FutilBackendHelper {
+
+  var idx: Map[String, Int] = Map();
+  def genName(base: String): CompVar = {
+    // update idx
+    idx get base match {
+      case Some(n) => idx = idx + (base -> (n + 1))
+      case None    => idx = idx + (base -> 0)
+    }
+    CompVar(s"$base${idx(base)}")
+  }
+
+  type Store = Map[CompVar, CompVar]
+
+  def emitDecl(d: Decl): List[Structure] = d.typ match {
+    case TArray(_, dims) => {
+      val const =
+        LibDecl(CompVar(s"${d.id}-init"), Stdlib.constant(VNone()))
+      val mem = LibDecl(CompVar(s"${d.id}"), Stdlib.memory(dims.map(_._1)))
+      List(
+        const,
+        mem,
+        Connect(const.id.port("out"), mem.id.port("data-in"))
+      )
+    }
+    case TBool() | TFloat() | TDouble() => {
+      val const =
+        LibDecl(CompVar(s"${d.id}-c"), Stdlib.constant(VNone()))
+      val reg = LibDecl(CompVar("${d.id}"), Stdlib.register())
+      List(const, reg, Connect(ThisPort(const.id), reg.id.port("in")))
+    }
+    case x => throw NotImplemented(s"Type $x not implemented for decls.")
+  }
+
+  def emitBinop(binop: CompInst, e1: Expr, e2: Expr)(
+      implicit store: Store
+  ): (Port, List[Structure], Enable) = {
+    val (e1port, e1struct, act1) = emitExpr(e1)
+    val (e2port, e2struct, act2) = emitExpr(e2)
+
+    val comp = LibDecl(genName(s"${binop.id}"), binop)
+    val struct = List(
+      comp,
+      Connect(e1port, comp.id.port("left")),
+      Connect(e2port, comp.id.port("right"))
+    )
+    (
+      comp.id.port("out"),
+      struct ++ e1struct ++ e2struct,
+      Enable(List(comp.id)) ++ act1 ++ act2
+    )
+  }
+
+  def emitExpr(expr: Expr, lhs: Boolean = false)(
+      implicit store: Store
+  ): (Port, List[Structure], Enable) =
+    expr match {
+      case EInt(v, _) => {
+        val const = LibDecl(genName("const"), Stdlib.constant(Num(v)))
+        (const.id.port("out"), List(const), Enable(List(const.id)))
+      }
+      case ERational(v: String) => {
+        val const = LibDecl(genName("const"), Stdlib.constant(v))
+        (const.id.port("out"), List(const), Enable(List(const.id)))
+      }
+      case EBinop(op, e1, e2) => {
+        op.op match {
```

Comment:
This case can be simplified to only get the string for `Stdlib.op` and then constructing the right Binop using that string.